velox/functions/lib/LikeFunctionUtils.cpp

```diff
@@ -0,0 +1,215 @@
+      return true;
+    }
+  }
+
+  return false;
+}
+
+// This function is called if the string 'pattern' has no wildcard characters; a
+// wildcard character could either be '_' or '%'. When there are no wildcard
+// characters in the pattern, the 'inputString' can be matched with the string
+// 'pattern' by comparing the characters at the same indices in 'inputString'
+// and 'pattern'. This character by character comparison is performed in this
+// function and if there is no mismatch till the end of both 'inputString' and
+// 'pattern' are reached, we return true.
+bool matchExactPattern(
+    const std::string& inputString,
+    const std::string& pattern) {
+  size_t inputStringIndex = 0;
+  size_t patternIndex = 0;
+  size_t inputStringLength = inputString.size();
+  size_t patternLength = pattern.size();
+
+  if (inputStringLength != patternLength) {
+    return false;
+  }
+
+  for (;
+       (inputStringIndex < inputStringLength) && (patternIndex < patternLength);
+       inputStringIndex++, patternIndex++) {
+    if (inputString[inputStringIndex] != pattern[patternIndex]) {
+      return false;
+    }
+  }
+  return (
+      (inputStringIndex == inputStringLength) &&
+      (patternIndex == patternLength));
+}
+
+// We use the Knuth-Morris-Pratt (KMP) string search algorithm to search for the
+// string 'pattern' in a given 'inputString'. More details about the KMP
+// algorithm can be found here:
+// https://www.geeksforgeeks.org/kmp-algorithm-for-pattern-searching/
+// https://cmps-people.ok.ubc.ca/ylucet/DS/KnuthMorrisPratt.html
+// This function does the preprocessing of the string 'pattern' to return the
+// array lps, which is used by the KMP algorithm. lps[i] stores the length of
+// the longest proper prefix that is also a suffix of the pat[0-i].
+void kmpSearchPreprocessPattern(
+    const std::string& pattern,
+    std::vector<int>& lps) {
+  size_t patternIndex = 1;
+  size_t lpsLength = 0;
+  size_t patternLength = pattern.size();
+
+  while (patternIndex < patternLength) {
+    if (pattern[patternIndex] == pattern[lpsLength]) {
+      lpsLength++;
+      lps[patternIndex] = lpsLength;
+      patternIndex++;
+    } else {
+      if (lpsLength != 0) {
+        lpsLength = lps[lpsLength - 1];
+      } else {
+        lps[patternIndex] = 0;
+        patternIndex++;
+      }
+    }
+  }
+}
+
+// This function checks if the string 'pattern' is an 'exact' pattern followed
+// by one or more occurences of the wildcard character, '%'; let us refer to
+// such patterns as prefix patterns. We define an 'exact' pattern as a string
+// that doesn't contain any wildcard characters; a wildcard character could
+// either be '_' or '%'. If 'pattern' is a prefix pattern, we return true.
+bool checkPrefixPattern(const std::string& pattern) {
+  size_t patternLength = pattern.size();
+  size_t patternIndex = patternLength - 1;
+
+  for (; patternIndex >= 0; patternIndex--) {
```

Comment:
Thanks for pointing out this case, it should be fine if these patterns are matched by isPrefix or isSuffix functions since they can be considered as either a prefix or suffix pattern, the functionality will not be affected. Should patterns with only wildcard characters be handled in a different function?

Reply:
There isn't a need for a separate function.

My understanding if the pattern is '%', then it should return true. However, from the logic I feel that pattern will hit the  check (patternIndex == patternLength - 1) and return false. Its important that isSuffix should also return false and then the regular path check the pattern '%'.

Reply:
Got it, thanks for clarifying, I checked the case where pattern is '%' and it is caught in the prefix pattern check, it doesn't hit the (patternIndex == patternLength - 1) check.